util/git/creds.go

```diff
@@ -33,21 +33,86 @@ func (c NopCreds) Environ() (io.Closer, []string, error) {
 
 // HTTPS creds implementation
 type HTTPSCreds struct {
+	// Username for authentication
 	username string
+	// Password for authentication
 	password string
+	// Whether to ignore invalid server certificates
 	insecure bool
+	// Client certificate to use
+	clientCertData string
+	// Client certificate key to use
+	clientCertKey string
 }
 
-func NewHTTPSCreds(username, password string, insecure bool) HTTPSCreds {
-	return HTTPSCreds{username, password, insecure}
+func NewHTTPSCreds(username, password, clientCertData, clientCertKey string, insecure bool) HTTPSCreds {
+	return HTTPSCreds{
+		username,
+		password,
+		insecure,
+		clientCertData,
+		clientCertKey,
+	}
 }
 
+// Get additional required environment variables for executing git client to
+// access specific repository via HTTPS.
 func (c HTTPSCreds) Environ() (io.Closer, []string, error) {
 	env := []string{fmt.Sprintf("GIT_ASKPASS=%s", "git-ask-pass.sh"), fmt.Sprintf("GIT_USERNAME=%s", c.username), fmt.Sprintf("GIT_PASSWORD=%s", c.password)}
+	httpCloser := authFilePaths(make([]string, 0))
+
+	// GIT_SSL_NO_VERIFY is used to tell git not to validate the server's cert at
+	// all.
 	if c.insecure {
 		env = append(env, "GIT_SSL_NO_VERIFY=true")
 	}
-	return NopCloser{}, env, nil
+
+	// In case the repo is configured for using a TLS client cert, we need to make
+	// sure git client will use it. The certificate's key must not be password
+	// protected.
+	if c.clientCertData != "" && c.clientCertKey != "" {
+		var certFile, keyFile *os.File
+
+		// We need to actually create two temp files, one for storing cert data and
+		// another for storing the key. If we fail to create second fail, the first
+		// must be removed.
+		certFile, err := ioutil.TempFile(util.TempDir, "")
+		if err == nil {
+			keyFile, err = ioutil.TempFile(util.TempDir, "")
```

Comment:
I feel like we should have a defer to delete the tmp files.

Reply:
Fixed.